metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/SampleEventHandler.java

```diff
@@ -31,16 +31,26 @@
     @HandleBeforeCreate
     @HandleBeforeSave
     public void validateTaxonomies(Sample sample) {
-        if (sample.getTaxonomies() == null || sample.getTaxonomies().size() == 0 || sample.getTaxonomies().get(0) == null) {
-            throw new IllegalArgumentException (ErrorMessage.SAMPLE_WITHOUT_TAXONOMY);
-        }
+        validateTaxonomyLink(sample);
     }
 
     @HandleBeforeLinkDelete
     @HandleBeforeLinkSave
     public void validateTaxonomies(Sample sample, Object taxonomy) {
-        if (sample.getTaxonomies() == null || sample.getTaxonomies().size() == 0 || sample.getTaxonomies().get(0) == null) {
+        validateTaxonomyLink(sample);
+    }
+
+    private void validateTaxonomyLink(Sample sample) {
+        if (sample.getTaxonomies() == null || sample.getTaxonomies().size() == 0) {
             throw new IllegalArgumentException (ErrorMessage.SAMPLE_WITHOUT_TAXONOMY);
+        } else if (sample.getTaxonomies().size() > 0) {
+            long invalid = sample.getTaxonomies()
```

Comment:
Stream the full array of links and counting is gonna have worse cost that just doing a for each either in a for or a stream.